src/main/scala/org/kframework/minikore/MiniKore.scala

```diff
@@ -1,39 +1,162 @@
 object MiniKore {
 
-  type Attributes = Seq[Pattern]
+  //Default Implementation
+
+  type Attributes = Seq[i.Pattern]
 
   case class Definition(modules: Seq[Module], att: Attributes)
+
   case class Module(name: String, sentences: Seq[Sentence], att: Attributes)
 
   sealed trait Sentence
+
   case class Import(name: String, att: Attributes) extends Sentence
+
   case class SortDeclaration(sort: String, att: Attributes) extends Sentence
+
   case class SymbolDeclaration(sort: String, label: String, args: Seq[String], att: Attributes) extends Sentence
-  case class Rule(pattern: Pattern, att: Attributes) extends Sentence
-  case class Axiom(pattern: Pattern, att: Attributes) extends Sentence
-
-  sealed trait Pattern
-  case class Variable(name: String, sort: String) extends Pattern
-  case class Application(label: String, args: Seq[Pattern]) extends Pattern
-  case class DomainValue(label: String, value: String) extends Pattern
-  //
-  case class True() extends Pattern
-  case class False() extends Pattern
-  //
-  case class And(p: Pattern, q: Pattern) extends Pattern
-  case class Or(p: Pattern, q: Pattern) extends Pattern
-  case class Not(p: Pattern) extends Pattern
-  case class Implies(p: Pattern, q: Pattern) extends Pattern
-  case class Exists(v: Variable, p: Pattern) extends Pattern
-  case class ForAll(v: Variable, p: Pattern) extends Pattern
-  //
-  case class Next(p: Pattern) extends Pattern
-  case class Rewrite(p: Pattern, q: Pattern) extends Pattern
-  case class Equal(p: Pattern, q: Pattern) extends Pattern
+
+  case class Rule(pattern: i.Pattern, att: Attributes) extends Sentence
+
+  case class Axiom(pattern: i.Pattern, att: Attributes) extends Sentence
+
+  case class Variable(name: String, sort: i.Sort) extends i.Variable {
+    override def build = DefaultBuilders.VariableBuilder
+  }
+
+  case class Application(label: String, args: Seq[i.Pattern]) extends i.Application {
+    override def build = DefaultBuilders.ApplicationBuilder(label, _)
+  }
+
+
+  case class DomainValue(label: String, value: String) extends i.DomainValue {
+    override def build = DefaultBuilders.DomainValueBuilder
+  }
+
+  case class True() extends i.True {
+    override def build = DefaultBuilders.TrueBuilder
+  }
+
+  case class False() extends i.False {
+    override def build: Node0Builder[i.Pattern] = DefaultBuilders.FalseBuilder
+  }
+
+  case class And(override val p: i.Pattern, override val q: i.Pattern) extends i.And {
+    override def build: Node2Builder[i.Pattern] = DefaultBuilders.AndBuilder
+  }
+
+  case class Or(p: i.Pattern, q: i.Pattern) extends i.Or {
+    override def build: Node2Builder[i.Pattern] = DefaultBuilders.OrBuilder
+  }
+
+  case class Not(p: i.Pattern) extends i.Not {
+    override def build: NodeBuilder[i.Pattern] = DefaultBuilders.NotBuilder
+  }
+
+
+  case class Implies(p: i.Pattern, q: i.Pattern) extends i.Implies {
+    override def build: Node2Builder[i.Pattern] = DefaultBuilders.ImpliesBuilder
```

Comment:
I would remove the type of `build` above, that is, the `: Node2Builder[i.Pattern]` part.  Scala can infer it on the one hand, and it makes the users be aware of "nodes" and the "ast" view.  We'd like to offer the "ast view for free" to people who are interested in it and appreciate its benefits; but we'd like the others to not have to pay any cost whatsoever, not even to have to be aware of it, if they chose to.